ethcontract-common/src/artifact.rs

```diff
@@ -99,3 +124,51 @@ impl Default for Artifact {
         Artifact::new()
     }
 }
+
+/// Result of inserting a nre contract into an artifact.
+pub struct InsertResult<'a> {
+    /// Reference to the newly inserted contract.
+    pub inserted_contract: ContractMut<'a>,
+
+    /// If insert operation replaced an old contract, it will appear here.
+    pub old_contract: Option<Contract>,
+}
+
+/// A wrapper that allows mutating contract
+/// but doesn't allow changing its name.
+pub struct ContractMut<'a>(&'a mut Contract);
```

Comment:
This is needed because modifying the contract name would cause the `Artifact`'s `HashMap<String, Contract>` to become inconsistent right?

Reply:
Exactly.